common/djangoapps/student/auth.py

```diff
@@ -230,9 +283,14 @@ def _check_caller_authority(caller, role):
     # superuser
     if GlobalStaff().has_user(caller):
         return
-
     if isinstance(role, (GlobalStaff, CourseCreatorRole, OrgContentCreatorRole)):  # lint-amnesty, pylint: disable=no-else-raise
         raise PermissionDenied
     elif isinstance(role, CourseRole):  # instructors can change the roles w/in their course
-        if not user_has_role(caller, CourseInstructorRole(role.course_key)):
+        # TODO: course roles: If the course roles feature flag is disabled the course_permission_check
+        # call below will never return true.
+        # Remove the user_has_role call when course_roles Django app are implemented.
+        if not (
+            user_has_role(caller, CourseInstructorRole(role.course_key)) or
+            course_permission_check(caller, CourseRolesPermission.MANAGE_ALL_USERS.value, role.course_key)
```

Comment:
I looked again at this code, trying to understand it at a deeper level. Specifically, if the intent behind a role/permissions solution is that all actions be gated by having appropriate permissions, then we should never have role-based tests. Put another way, if a role is duplicated under a new name, then the original role and the duplicate will have the same permissions. A role-based tests would pass the first role but fail the duplicate, whereas a permissions-based test would pass both. Clearly the latter is what should happen, and this corroborates the notion that gates should always be permission-based, and not role-based. What this example shows is that gates in existing code are sometimes role-based, and when this occurs it's necessary to convert them to permission-based gates.

Part of the complication in doing this is that we may need to generate new permissions to deal with such cases. Here, do we need a CourseRolesPermission.MANAGE_ROLES permission? It strikes me that the ability to modify which permissions go with a role is outside what I would expect in a MANAGE_ALL_USERS permission. MANAGE_ROLES is a permission a platform admin role might have, whereas MANAGE_ALL_USERS is a permission a course owner role might have.

Reply:
The goal of this PR is to convert role based gating to a combined role and permission based gating as an intermediate step to be used until all existing course level roles (student_courseaccessrole and django_commentclientroles) are migrated over to course_roles_role. 

We worked with product to determine the initial list of permissions, but a permission was not created that would allow a user to modify which permissions belong to a role. CourseRoles was designed to allow for adding permissions in the future (with code changes) and roles in the future (without needing to modify code). This full design is not implemented in this PR, but this sets the foundational work for that end goal. 

To my knowledge there are not current plans to make modifying the permissions assigned to a role (or even creating a new role) something that is possible within the LMS/CMS. If this changes, we will need a new permission to address that need, but it likely would not be a course level permission, but rather an instance or org level permission. 

From tracing this code usage of this function, we determined it was used to decide who could assign roles to users which fit the MANAGE_ALL_USERS permission. If we mis-interpreted the code usage, please let us know.

Reply:
Course staff are not given the MANAGE_ALL_USERS permission, so this test will only allow the course instructor to make these changes, as is currently the case. But when I look at the `user_has_role()` test, I immediately understand what's intended. When I look at the `MANAGE_ALL_USERS` test, the intent is far less clear. I understand your reasoning, but it seems like a step backwards in terms of code clarity.

Reply:
We will be changing the function names to increase clarity on their purpose.